crates/derive/src/stages/attributes_queue/builder.rs

```diff
@@ -0,0 +1,170 @@
+//! The [`AttributesBuilder`] and it's default implementation.
+
+use crate::{
+    params::SEQUENCER_FEE_VAULT_ADDRESS,
+    types::{
+        BlockID, BlockInfo, BuilderError, L2BlockInfo, PayloadAttributes, RawTransaction, Receipt,
+        RollupConfig, SystemConfig,
+    },
+};
+use alloc::{fmt::Debug, sync::Arc, vec, vec::Vec};
+use alloy_primitives::B256;
+
+/// The [AttributesBuilder] is responsible for preparing [PayloadAttributes]
+/// that can be used to construct an L2 Block containing only deposits.
+pub trait AttributesBuilder {
+    /// Prepares a template [PayloadAttributes] that is ready to be used to build an L2 block.
+    /// The block will contain deposits only, on top of the given L2 parent, with the L1 origin
+    /// set to the given epoch.
+    /// By default, the [PayloadAttributes] template will have `no_tx_pool` set to true,
+    /// and no sequencer transactions. The caller has to modify the template to add transactions.
+    /// This can be done by either setting the `no_tx_pool` to false as sequencer, or by appending
+    /// batch transactions as the verifier.
+    fn prepare_payload_attributes(
+        &mut self,
+        l2_parent: L2BlockInfo,
+        epoch: BlockID,
+    ) -> Result<PayloadAttributes, BuilderError>;
+}
+
+/// The [L1ReceiptsFetcher] fetches L1 Header Info and [Receipt]s for
```

Comment:
We should add a new function to `ChainProvider` for `info_by_hash` - we already have `receipts_by_hash` in there 😄